File: masar_mce/custom/purchase_order/purchase_order.js
frappe.ui.form.on("Purchase Order", {
    setup(frm) {
        FilterItems(frm);
        CreatePurchaseRequest(frm);
    }, 
    refresh(frm) {
        FilterItems(frm);
        CreatePurchaseRequest(frm);
    }, 
    onload(frm) {
        FilterItems(frm);
        CreatePurchaseRequest(frm);
    }, 
    supplier(frm){
        FilterItems(frm);
    }
});

function FilterItems(frm) {
    const grid = frm.fields_dict.items.grid;
    const item_code_field = grid.get_field("item_code");
    if (!item_code_field.hasOwnProperty('original_get_query')) {
        item_code_field.original_get_query = item_code_field.get_query;
    }
    
    item_code_field.get_query = function() {
        if (frm.doc.supplier) {
            return {
                query: "masar_mce.custom.purchase_order.purchase_order.get_items_from_active_blanket_order",
                filters: {
                    supplier: frm.doc.supplier
                }
            };
        } else {
            return item_code_field.original_get_query ? item_code_field.original_get_query() : {};
        }
    };
    setTimeout(() => {    
            frm.remove_custom_button("Link to Material Request", "Tools");
            frm.remove_custom_button("Update Rate as per Last Purchase", "Tools");
            cur_frm.page.remove_inner_button(__('Payment'),  __('Create'));
            cur_frm.page.remove_inner_button(__('Payment Request'),  __('Create'));
            cur_frm.page.remove_inner_button(__('Purchase Invoice'),  __('Create'));
            cur_frm.page.remove_inner_button(__('Product Bundle'),  __('Get Items From'));
            cur_frm.page.remove_inner_button(__('Material Request'),  __('Get Items From'));
            cur_frm.page.remove_inner_button(__('Supplier Quotation'),  __('Get Items From'));
        },100);
}
frappe.ui.form.on('Purchase Order Item', {
    item_code: function(frm, cdt, cdn) {
        GetItemDetails(frm , cdt , cdn)
    } , 
    rate :  function(frm, cdt, cdn) {
        GetItemDetails(frm , cdt , cdn)
    } , 
});
function GetItemDetails(frm , cdt , cdn){
    const row = locals[cdt][cdn];
        if (!frm.doc.supplier || !row.item_code) {
            return;
        }
        frappe.call({
            method: "masar_mce.custom.purchase_order.purchase_order.get_blanket_order_for_item",
            args: {
                supplier: frm.doc.supplier,
                item_code: row.item_code
            },
            callback: function(r) {
                if (r.message) {
                    const data = r.message;
                    frappe.model.set_value(cdt, cdn, 'against_blanket_order' , 1);
                    frappe.model.set_value(cdt, cdn, 'blanket_order', data.parent);
                    frappe.model.set_value(cdt, cdn,'custom_blanket_order_item', data.name ) ;
                    frappe.model.set_value(cdt, cdn, 'rate', data.rate);
                }
            }
        });
}
frappe.form.link_formatters['Item'] = function(value, doc) {
    if(doc.item_code && doc.item_name !== value) {
        return doc.item_code;
    } else {
        return value;
    }
};
function CreatePurchaseRequest(frm) {
    if (frm.doc.docstatus === 1 ) {
            frm.add_custom_button(__('Purchase Request'), function() {
                frappe.model.open_mapped_doc({
                    method: "masar_mce.custom.purchase_order.purchase_order.create_purchase_request_from_purchase_order",
                    frm: cur_frm,
                    freeze_message: __("Creating Purchase Request ..."),
                });
            }, __('Create'));
        }
}

File: masar_mce/custom/purchase_order/purchase_order.py
import frappe , json
from frappe.utils import flt ,date_diff, add_days
from frappe.model.mapper import get_mapped_doc
from erpnext.buying.doctype.purchase_order.purchase_order import set_missing_values
@frappe.whitelist()
def get_blanket_order_for_item(item_code, supplier):
    if not item_code or not supplier:
        return {}
    result = frappe.db.sql(
        """
        SELECT bo.name AS parent, boi.name AS name , boi.rate as rate
        FROM `tabBlanket Order` bo
        INNER JOIN `tabBlanket Order Item` boi ON bo.name = boi.parent
        WHERE bo.supplier = %(supplier)s
          AND bo.docstatus = 1
          AND bo.custom_status = 'Active'
          AND boi.item_code = %(item_code)s
        LIMIT 1
        """,
        {"supplier": supplier, "item_code": item_code},
        as_dict=True
    )
    return result[0] if result else {}


@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def get_items_from_active_blanket_order(doctype, txt, searchfield, start, page_len, filters):
    supplier = filters.get("supplier")
    if not supplier:
        return []
    query = """
        SELECT DISTINCT
            boi.item_code,
            item.item_name
        FROM `tabBlanket Order` bo
        INNER JOIN `tabBlanket Order Item` boi ON bo.name = boi.parent
        INNER JOIN `tabItem` item ON boi.item_code = item.name
        WHERE bo.supplier = %(supplier)s
        AND bo.docstatus = 1
        AND bo.custom_status = 'Active'
        AND item.disabled = 0
        AND (boi.item_code LIKE %(txt)s OR item.item_name LIKE %(txt)s)
        ORDER BY boi.item_code
        LIMIT %(start)s, %(page_len)s
    """
    return frappe.db.sql(query, {
        "supplier": supplier,
        "txt": f"%{txt}%",
        "start": start,
        "page_len": page_len
    })


@frappe.whitelist()
def create_purchase_request_from_purchase_order(source_name, target_doc=None, args=None):
    if args is None:
        args = {}
    if isinstance(args, str):
        args = json.loads(args)
    def update_item(obj, target, source_parent):
        target.uom = obj.uom
        target.item_code = obj.item_code
        target.description = obj.description
        target.stock_uom = obj.stock_uom
        target.request_quantity = flt(obj.qty) - flt(obj.received_qty)
        target.purchase_order = obj.parent
        target.purchase_order_item = obj.name
    def select_item(d):
        filtered_items = args.get("filtered_children", [])
        child_filter = d.name in filtered_items if filtered_items else True
        return child_filter

    doc = get_mapped_doc(
        "Purchase Order",
        source_name,
        {
            "Purchase Order": {
                "doctype": "Market Purchase Request",
                "field_map": {
                    "supplier": "supplier", 
                },
                "validation": {
                    "docstatus": ["=", 1],
                },
            },
            "Purchase Order Item": {
                "doctype": "Purchase Request Item",
                "field_map": {
                    "name": "purchase_order_item",
                    "parent": "purchase_order",
                },
                "postprocess": update_item,
                "condition": lambda doc: select_item(doc),
            },
        },
        target_doc
    )

    return doc
    
def on_submit(self , method):
    set_agreement_date(self)
    

def set_agreement_date(self):
    agreements = frappe.db.sql("""
        SELECT DISTINCT poi.blanket_order
        FROM `tabPurchase Order Item` poi
        WHERE poi.parent = %s
          AND poi.blanket_order IS NOT NULL
    """, self.name, as_dict=1)

    for agr in agreements:
        if not agr.blanket_order:
            continue

        agr_doc = frappe.get_doc("Blanket Order", agr.blanket_order)

        if not agr_doc.custom_posting_date:
            diff_days = date_diff(self.transaction_date, agr_doc.from_date)

            new_from_date = add_days(agr_doc.from_date, diff_days)
            new_to_date   = add_days(agr_doc.to_date, diff_days)

            frappe.db.set_value("Blanket Order", agr.blanket_order, {
                "custom_posting_date": agr_doc.from_date,
                "from_date": new_from_date,
                "to_date": new_to_date
            })
            
def before_insert(self, method):

    item_codes = [d.item_code for d in self.items]

    prices = frappe.db.sql("""
        SELECT 
            ip.item_code,
            ip.price_list_rate
        FROM `tabItem Price` ip
        INNER JOIN (
            SELECT 
                item_code,
                MAX(valid_from) AS latest_date
            FROM `tabItem Price`
            WHERE buying = 1
              AND item_code IN %(items)s
            GROUP BY item_code
        ) latest
            ON latest.item_code = ip.item_code
           AND latest.latest_date = ip.valid_from
        WHERE ip.buying = 1
    """, {"items": tuple(item_codes)}, as_dict=1)

    price_map = {p.item_code: p.price_list_rate for p in prices}

    for i in self.items:
        i.rate = price_map.get(i.item_code, 0)